Add alphabetical sorting to Animation Player

Arnklit opened the pull request:
This adds alphabetical sorting to the Animation Editor. This might close https://github.com/godotengine/godot-proposals/issues/3426.

https://github.com/user-attachments/assets/45d21e7a-cff1-48e4-895f-670457c820b9

Arnklit committed:
Add alphabetical sorting to Animation Player

Arnklit commented on editor/animation_track_editor.cpp:
```diff
@@ -7653,6 +7683,14 @@ AnimationTrackEditor::AnimationTrackEditor() {
+	alphabetic_sorting->set_flat(true);
+	alphabetic_sorting->connect(SceneStringName(pressed), callable_mp(this, &AnimationTrackEditor::_sort_alphabetically_toggle));
+	alphabetic_sorting->set_toggle_mode(true);
+	alphabetic_sorting->set_tooltip_text(TTR("Sort tracks/groups alphabetically."));
```
Fixed

Arnklit commented:
I've slightly restructured the code with some help from @mihe as it felt a bit hacky before. Looks really good not I think.

arkology commented on editor/animation_track_editor.cpp:
```diff
@@ -7653,6 +7696,14 @@ AnimationTrackEditor::AnimationTrackEditor() {
 
+	alphabetic_sorting = memnew(Button);
+	alphabetic_sorting->set_flat(true);
+	alphabetic_sorting->connect(SceneStringName(pressed), callable_mp(this, &AnimationTrackEditor::_sort_alphabetically_toggle));
```
I wonder if it makes sense to add a new function (`_sort_alphabetically_toggle()`) to call already existing function  (`_update_tracks();`). Maybe connect it directly to `_update_tracks`?

Arnklit replied:
Yes good point. When I set this up I think I imagined more logic would go into that function. Removed it and connected directly to `_update_tracks()`

Calinou commented on editor/animation_track_editor.cpp:
```diff
@@ -7653,6 +7692,14 @@ AnimationTrackEditor::AnimationTrackEditor() {
+	alphabetic_sorting->set_flat(true);
+	alphabetic_sorting->connect(SceneStringName(pressed), callable_mp(this, &AnimationTrackEditor::_update_tracks));
+	alphabetic_sorting->set_toggle_mode(true);
+	alphabetic_sorting->set_tooltip_text(TTR("Sort tracks/groups alphabetically.\nIf disabled, tracks are shown in the order they are added and can be reordered using drag-and-drop."));
```
This doesn't appear to be true after testing it. I was convinced it was possible to drag-and-drop to reorder though, so I remember it wrong.

```suggestion
	alphabetic_sorting->set_tooltip_text(TTR("Sort tracks/groups alphabetically.\nIf disabled, tracks are shown in the order they are added."));
```

Arnklit replied:
I don't think this is right. You can reorder tracks, just only within the group when in grouped mode, which makes sense. I made a change so you can do this when the groups are sorted alphabetically as well, but of course it makes no sense to allow in tracks only view.

Calinou replied:
I see… What I was trying to achieve was to reorder groups around.

Calinou reviewed (approved):
Tested locally, it works as expected. Code looks good to me.

Note that the option is global (it applies to all scenes in the current project) and does not persist across editor restarts. This is probably good to go as-is, I'm pointing this out just in case.

Arnklit commented on editor/animation_track_editor.h:
```diff
@@ -943,6 +952,7 @@ class AnimationTrackEditor : public VBoxContainer {
 	float snap_time(float p_value, bool p_relative = false);
 	float get_snap_unit();
 	bool is_grouping_tracks();
+	bool is_alphabetically_sorting();
```
fixed

Arnklit commented on editor/animation_track_editor.cpp:
```diff
@@ -4850,6 +4850,7 @@ void AnimationTrackEditor::_update_tracks() {
 
 	bool use_grouping = !view_group->is_pressed();
 	bool use_filter = selected_filter->is_pressed();
+	bool use_alphabetical_sorting = alphabetic_sorting->is_pressed();
```
fixed

Arnklit commented on editor/animation_track_editor.cpp:
```diff
@@ -5000,6 +5003,37 @@ void AnimationTrackEditor::_update_tracks() {
+		if (use_alphabetical_sorting) {
+			struct TrackAlphaCompare {
+				bool operator()(const AnimationTrackEdit *a, const AnimationTrackEdit *b) const {
+					return (String)a->get_path() < (String)b->get_path();
```
fixed

SaracenOne reviewed (approved):
LGTM 👍

Arnklit commented:
Set to draft. Found some issues with this in testing. The path used for sorting is not correct (it's using the full path, rather than the leaf) and it was not as easy to solve as first assumed (especially with regards to groups).

Arnklit commented:
Rebased and fixed the issues with the sorting being based on the whole path rather than just the leaf. It meant I had to sort them in group view as well, rather than just relying on the order from the RedBlack set. To sort them I needed to expose the node_name on AnimationTrackEditGroup, so I made a getter for that.

Calinou reviewed (approved):
Tested locally, it works as expected. Code looks good to me.

Repiteo commented:
Thanks!
